proposals/exception-handling/Exceptions-formal-overview.md

```diff
@@ -0,0 +1,166 @@
+# 3rd proposal formal spec overview
+
+This is an overview of the 3rd proposal's formal spec additions, to aid in discussions concerning the proposed semantics.
+
+## Abstract Syntax
+
+### Types
+
+Exception Types
+
+```
+exntype ::= [t*] -> []
+```
+
+### Instructions
+
+```
+instr ::= ... | throw x | rethrow l
+        | try bt instr* (catch x instr*)* (catch_all instr*)? end
+        | try bt instr* delegate l
+        | try bt instr* unwind instr* end
+```
+
+### Modules
+
+Exceptions (definitions)
+
+```
+exn ::= export* exception exntype  | export* exception exntype import
+```
+
+Modules
+
+
+```
+mod ::= module ... exn*
+```
+
+## Validation (Typing)
+
+To verify that the `rethrow l` instruction refers to a surrounding catch block, we introduce a stack `caught` to validation contexts, which gets an exception index or the keywork `all` prepended whenever we enter instructions inside a `catch exnidx` or `catch_all` block, respectively. This addition is reflected in the execution rules, by the administrative instruction `caught` which models the stack of caught exceptions on the wasm stack.
+
+
+### Instructions
+
+
+
+```
+C_exn(x) = [t*] -> []
+--------------------------------
+C |- throw x : [t1* t*] -> [t2*]
+
+
+C_caught(l) =/= ε
+-------------------------------
+C |- rethrow l : [t1*] -> [t2*]
+
+
+bt = [t1*] -> [t2*]
+C, label [t2*] |- instr* : [t1*] -> [t2*]
+(C_exn(x_i) = [t'_i*] -> [])^n
+(C, label [t2*], caught x_i |- instr_i* : [t'_i*] -> [t2*])^n
+(C, label [t2*], caught all |- instr'* : [] -> [t2*])?
+------------------------------------------------------------------
+try bt instr* (catch x_i instr_i*)^n (catch_all instr'*)? end : bt
+
+
+bt = [t1*] -> [t2*]
+C, label [t2*] |- instr* : [t1*] -> [t2*]
+C_label(l) =/= ε
+-----------------------------------------
+try bt instr* delegate l : bt
+
+
+bt = [t1*] -> [t2*]
+C, label [t2*] |- instr_1* : [t1*] -> [t2*]
+C, label [t2*], caught all |- instr_2* : [] -> [t2*]
+----------------------------------------------------
+try bt instr_1* unwind instr_2* : bt
```

Comment:
```suggestion
C, label [t2*] |- instr_2* : [] -> []
----------------------------------------------------
try bt instr_1* unwind instr_2* end : bt
```

This was discussed in #142, but just to reiterate for the sake of having this suggestion be self contained: A purpose of `unwind` is to specify unwinders in a way that works well with non-exceptional non-local control constructs. To that end `rethrow` should not work within `unwind` (which was the case in the earlier formalization), as it may not make sense to `rethrow` other non-local control constructs from within a different dynamic scope. Instead, `unwind` automatically resumes unwinding at the `end`, and so `instr_2*` does not need to output `[t2*]`.

Reply:
Issue #142 is still open, and (unfortunately) it does not seem to me that there is general agreement on `unwind`. About the return values, in the closed issue #129, it was decided that unwind blocks should indeed return values.

About `rethrow` in `unwind` blocks or not, I can't find that discussion, can you please link to it?

By the way, the earlier formalisation was wrong, as it did not reflect the spec that @aheejin presented and that we voted on.

Reply:
Thank you for catching the missing `end` typo!

Reply:
@RossTate, about the `unwind` block's instructions returning values or not: now I see that I was perhaps too quick to close #129, since you don't seem to agree with @aheejin's [proposal there](https://github.com/WebAssembly/exception-handling/issues/129#issuecomment-705800164). Since `unwind` ends with a `rethrow 0`, any return values seem indeed unnecessary. @aheejin what do you think?

Reply:
@RossTate about `rethrow` inside an `unwind` block's `instr_2*`, I can't see much difference from having a `throw` inside `instr_2*`. Looking around the issues, I only saw arguments saying that a `rethrow` would end the unwinding process, but so would a `br` or `throw` or `return`. I didn't manage to find a discussion wrt `rethrow` of other non-local control constructs. Please help me out find the relevant comments/discussions, I'm a bit lost in the several issues containing mixed information. 

Having said that, please note that the `rethrow 0` resulting from the current `unwind`, is creating a new exception, so _it's throwing within a different dynamic scope,_ IIUC. I do understand that `unwind` should probably be laid out more carefully than what we do here, and this is exactly why I suggested that we postpone it to an extension of the MVP, to avoid making mistakes we can't take back anymore, wrt backwards compatibility. I don't think that people should expect the MVP to be WebAssembly's "final word on exceptions" (or non-local control flow in general - as @phoe suggested), but the "first word on exceptions" (or non-local control flow). :) 

But, as @aheejin  says, I'm also ok to keep `unwind`, if it helps us get the spec settled soon so we can continue our implementations. (Btw, @takikawa and I already started landing first patches for `try-catch` and `throw` in [SpiderMonkey](https://bugzilla.mozilla.org/show_bug.cgi?id=1335652).)

Reply:
@ioannad 
> @RossTate, about the `unwind` block's instructions returning values or not: now I see that I was perhaps too quick to close #129, since you don't seem to agree with @aheejin's [proposal there](https://github.com/WebAssembly/exception-handling/issues/129#issuecomment-705800164). Since `unwind` ends with a `rethrow 0`, any return values seem indeed unnecessary to me as well. @aheejin what do you think? The try block can still have any blocktype, it's just that the unwind block's `instr_2` shouldn't return any values, since they will be discarded by the `rethrow` in the end anyway.

Sorry I don't understand. Throwing values are different from returning values, and they don't have to match. we can branch out of `unwind` block so that we don't end up rethrowing anything, in which case return values make sense, no?

> Having said that, please note that the `rethrow 0` resulting from the current `unwind`, is creating a new exception, so _it's throwing within a different dynamic scope,_ IIUC. 

I thought `rethrow` does not create a new exception, and neither does `end` at the `unwind`..? Why does it create a new exception? If it does, does it preserve metadata like stack traces? I think it should.

Reply:
@aheejin 

> > @RossTate, about the `unwind` block's instructions returning values or not: now I see that I was perhaps too quick to close #129, since you don't seem to agree with @aheejin's [proposal there](https://github.com/WebAssembly/exception-handling/issues/129#issuecomment-705800164). Since `unwind` ends with a `rethrow 0`, any return values seem indeed unnecessary to me as well. @aheejin what do you think? The try block can still have any blocktype, it's just that the unwind block's `instr_2` shouldn't return any values, since they will be discarded by the `rethrow` in the end anyway.
> 
> Sorry I don't understand. Throwing values are different from returning values, and they don't have to match. we can branch out of `unwind` block so that we don't end up rethrowing anything, in which case return values make sense, no?

I'm not sure I understand. Wouldn't that be branching from "the middle" of the unwind block? I think in that case the part of the unwind block's instructions before the branch should match the branch destination. Then depending on whether we allow return values from unwind blocks or not, we could drop them or adjust them before the end of the unwind block's instructions. Does that make sense?

> 
> > Having said that, please note that the `rethrow 0` resulting from the current `unwind`, is creating a new exception, so _it's throwing within a different dynamic scope,_ IIUC.
> 
> I thought `rethrow` does not create a new exception, and neither does `end` at the `unwind`..? Why does it create a new exception? If it does, does it preserve metadata like stack traces? I think it should.

You're right of course, what I should have said is that `rethrow` is taking an exception possibly originating elsewhere, and throwing it from the context of the `unwind` instruction. So it's not a new exception, but a new throw: rethrow is not propagating the exception, it's throwing it from the location of the rethrow instruction. Does this match your understanding?

Reply:
> @aheejin
> 
> > > @RossTate, about the `unwind` block's instructions returning values or not: now I see that I was perhaps too quick to close #129, since you don't seem to agree with @aheejin's [proposal there](https://github.com/WebAssembly/exception-handling/issues/129#issuecomment-705800164). Since `unwind` ends with a `rethrow 0`, any return values seem indeed unnecessary to me as well. @aheejin what do you think? The try block can still have any blocktype, it's just that the unwind block's `instr_2` shouldn't return any values, since they will be discarded by the `rethrow` in the end anyway.
> > 
> > 
> > Sorry I don't understand. Throwing values are different from returning values, and they don't have to match. we can branch out of `unwind` block so that we don't end up rethrowing anything, in which case return values make sense, no?
> 
> I'm not sure I understand. Wouldn't that be branching from "the middle" of the unwind block? I think in that case the part of the unwind block's instructions before the branch should match the branch destination. Then depending on whether we allow return values from unwind blocks or not, we could drop them or adjust them before the end of the unwind block's instructions. Does that make sense?

I'm not sure why branching from the middle of the unwind block is not allowed? Branching from the middle of a try or a catch block is allowed too, and the return type of those don't have to match the exception type any instruction can throw within the try block. Perhaps I'm not understanding what you are explaining. Can you provide an example?

> 
> > > Having said that, please note that the `rethrow 0` resulting from the current `unwind`, is creating a new exception, so _it's throwing within a different dynamic scope,_ IIUC.
> > 
> > 
> > I thought `rethrow` does not create a new exception, and neither does `end` at the `unwind`..? Why does it create a new exception? If it does, does it preserve metadata like stack traces? I think it should.
> 
> You're right of course, what I should have said is that `rethrow` is taking an exception possibly originating elsewhere, and throwing it from the context of the `unwind` instruction. So it's not a new exception, but a new throw: rethrow is not propagating the exception, it's throwing it from the location of the rethrow instruction. Does this match your understanding?

Yes, I think that makes sense.

Reply:
Sorry to bump this old thread, but I'm not sure a conclusion was reached on the blocktype issue for `unwind`. I'd like to try to clarify with some examples in this thread as all the context is here (but I could open a new issue if that's preferred).

Here are 3 small `unwind` examples. I was wondering which of them should validate:

```
;; Example 1
try (result i32)
  call $f
unwind
  (i32.const 42) ;; matches block type, but value can never be used
end

;; Example 2
try (result i32)
  call $f
unwind
  nop ;; missing i32
end

;; Example 3
try (result)
  call $f
unwind
  (i32.const 42) ;; too many i32s
end
```

My understanding based on the discussion here and the current explainer & spec was that only example 1 should validate, and that examples 2 and 3 are invalid. (in all examples, the stack will be reset by the implicit rethrow of `end` so the values do not matter, but only in 1 is it consistent with the block type)

Does this sound right? (I'm asking for clarification because I think I have implemented this validation slightly differently, so I will go fix those if there's agreement on this)

Reply:
My understanding is also that only the example 1 should validate. Let us know if others have other opinions.

Reply:
I agree that all unwind blocks should be typed `[] -> []`. That's independent of the try's block type, since its a completely separate execution path. So, Example 2 is the only one that ought to type-check.

Reply:
@RossTate What's your take on the typing rule of `unwind` block?

Reply:
Assuming the `[] -> []` type, what is the implication for something like a `br 0` inside the unwind block? Since it exits the block, it needs to leave values on the stack that match the try's signature, so it would be validated using the try block's type rather than the enclosing unwind block's type?

Reply:
@thibaudmichaud, the br is typed according to the label type, which would still be `t2*`.

Reply:
If we can place `br` that returns a value within `unwind`, I'm not sure why the type of `unwind` block has to be `[] -> []`?

Reply:
`br` and the block's `end` are control paths with different targets in the case of unwind, so they have different type constraints (that's also the case with `loop` btw). An unwind block will drop any remaining value on the floor at `end`, so it's pointless to require any values there -- all that would do is force producers to push dummy values for no reason.

Reply:
Sorry I'm late to the party here, but I would expect all three of @takikawa's examples to validate because of implicit unreachable polymorphic behavior due to the implicit `rethrow` at the end of the `unwind` clause, just as they would if the `rethrow` were made explicit as a separate instruction at the end of the block.

Reply:
@tlively, that would be possible as well, but is useless in this case, since unlike branches or throws, an `end` cannot occur in the middle of a larger "expression". Stack polymorphism is useful for examples like
```
(... larger exp (i32.div_u (x) (br_if $bailout (i32.eqz (y))) (y)) ...)
```
or something like that. That cannot be constructed with a delimiter like `end`.

Not saying that we can't do it, but it would be a (minor) relaxation without practical value.

Reply:
I agree there’s no practical value. Not making that relaxation would be fine with me.